---
title: "Importing scRNA-seq Data for Commandline QC"
---

Single Cell ToolKit (SingleCellTK, SCTK) enables importing data from the following pre-processing tools-

- [CellRanger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger)
- [Optimus](https://data.humancellatlas.org/pipelines/optimus-workflow)
- [DropEst](https://github.com/hms-dbmi/dropEst)
- [BUStools](https://github.com/BUStools/bustools)
- [Seqc](https://github.com/ambrosejcarr/seqc)
- [STARSolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md)
- [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) object saved in RDS file
- [AnnData](https://github.com/theislab/anndata) object saved in  HDF5 file

The following table describes how SCTK expects the inputs to be structured and passed for each import function. In all cases, SCTK retains the standard output directory structure from upstream tools. All the import functions return the imported counts matrix as an `assay` in a `SingleCellExperiment` object, with associated information in respective `colData`, `rowData`, `reducedDims`, and `metadata` fields. 

![QC data format](qc_inputShell.png)\

Detailed information on function parameters and defaults are available in the [*Reference*](../../reference/index.html) section. Please also refer to [*Commandline Quality Control* documentation](cmd_qc.html#parameters-1) for the detail of other commandline input arguments.
